Compilation Error on `pyston_master` branch (GNU/Linux, gcc 8.3.0, python 3.7.3)

I attempted to compile the code on the `pyston_master` branch, installed dependencies, ran `../configure --enable-optimizations` and then ran `make`, and got the following compilation error:

```
gcc -pthread -c -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -DENABLE_AOT   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -fprofile-generate -I../Include/internal -IObjects -IInclude -IPython -I. -I../Include   -I../../aot  -DPy_BUILD_CORE -o Python/import.o ../Python/import.c
../Python/import.c: In function ‘_PyImport_Fini2’:
../Python/import.c:300:5: error: implicit declaration of function ‘_PyMem_SetDefaultAllocator’; did you mean ‘_PyMem_SetupAllocators’? [-Werror=implicit-function-declaration]
     _PyMem_SetDefaultAllocator(PYMEM_DOMAIN_RAW, &old_alloc);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~
     _PyMem_SetupAllocators
cc1: some warnings being treated as errors
```

When I try using the latest release, I get the same exact error.

I currently have Python 3.7.3 installed; do I simply need to install python 3.8?

Thanks very much-! A plain old `make` invocation did progress beyond the previous error. However, now I'm getting a new error with the following:
```
/tmp/tmp_env/bin/python3 pyston/run_profile_task.py
Traceback (most recent call last):
  File "pyston/run_profile_task.py", line 13, in <module>
    run("macrobenchmarks/benchmarks/djangocms.py", "800") # approx 21s
  File "pyston/run_profile_task.py", line 8, in run
    subprocess.check_call(["/usr/bin/time", exe, bench] + list(args))
  File "/home/chozorho/git/pyston/pyston/build/cpython_opt_install/usr/lib/pyston3.8/subprocess.py", line 359, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/home/chozorho/git/pyston/pyston/build/cpython_opt_install/usr/lib/pyston3.8/subprocess.py", line 340, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/home/chozorho/git/pyston/pyston/build/cpython_opt_install/usr/lib/pyston3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/home/chozorho/git/pyston/pyston/build/cpython_opt_install/usr/lib/pyston3.8/subprocess.py", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/usr/bin/time'
make: *** [Makefile:190: pyston/build/cpython_opt_install/usr/bin/python3.bolt.fdata] Error 1
```

I know that `time` is a command I can invoke (I use it from the terminal every once in a while to measure command runtimes). But if I type `locate time`, nothing comes up, so I have no idea where this utility is installed, but it's definitely not in `/usr/bin`.

@undingen Do you have any advice for me? And if not, is it worth opening a new Issue for it?

--- a/Include/internal/pycore_pymem.h
+++ b/Include/internal/pycore_pymem.h
@@ -148,7 +148,7 @@ struct _gc_runtime_state {
 PyAPI_FUNC(void) _PyGC_Initialize(struct _gc_runtime_state *);
 
 
-#if PY_DEBUGGING_HOOKS
+#if PY_DEBUGGING_FEATURES
 /* Set the memory allocator of the specified domain to the default.
    Save the old allocator into *old_alloc if it's non-NULL.
    Return on success, or return -1 if the domain is unknown. */
@@ -171,7 +171,7 @@ PyAPI_FUNC(int) _PyMem_SetDefaultAllocator(
 #define PYMEM_CLEANBYTE      0xCD
 #define PYMEM_DEADBYTE       0xDD
 #define PYMEM_FORBIDDENBYTE  0xFD
-#if PY_DEBUGGING_HOOKS
+#if PY_DEBUGGING_FEATURES
 
 /* Heuristic checking if a pointer value is newly allocated
    (uninitialized), newly freed or NULL (is equal to zero).

--- a/Makefile
+++ b/Makefile
@@ -77,27 +77,27 @@ PROFILE_TASK:=../../../Lib/test/regrtest.py -j 1 -unone,decimal -x test_posix te
 MAKEFILE_DEPENDENCIES:=Makefile.pre.in configure
 pyston/build/cpython_bc_build/Makefile: $(CLANG) $(MAKEFILE_DEPENDENCIES)
 	mkdir -p pyston/build/cpython_bc_build
-	cd pyston/build/cpython_bc_build; WRAPPER_REALCC=$(realpath $(CLANG)) WRAPPER_OUTPUT_PREFIX=../cpython_bc CC=../../../pyston/clang_wrapper.py CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -Wno-unused-command-line-argument" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --disable-aot --disable-debugging-features
+	cd pyston/build/cpython_bc_build; WRAPPER_REALCC=$(realpath $(CLANG)) WRAPPER_OUTPUT_PREFIX=../cpython_bc CC=../../../pyston/clang_wrapper.py CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -Wno-unused-command-line-argument" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --disable-aot --disable-debugging-features --enable-configure
 pyston/build/cpython_unopt_build/Makefile: $(MAKEFILE_DEPENDENCIES)
 	mkdir -p pyston/build/cpython_unopt_build
-	cd pyston/build/cpython_unopt_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -fno-reorder-blocks-and-partition" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS) -Wl,--emit-relocs" ../../../configure --prefix=/usr --disable-debugging-features
+	cd pyston/build/cpython_unopt_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -fno-reorder-blocks-and-partition" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS) -Wl,--emit-relocs" ../../../configure --prefix=/usr --disable-debugging-features --enable-configure
 pyston/build/cpython_opt_build/Makefile: $(MAKEFILE_DEPENDENCIES)
 	mkdir -p pyston/build/cpython_opt_build
-	cd pyston/build/cpython_opt_build; PROFILE_TASK="$(PROFILE_TASK)" CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -fno-reorder-blocks-and-partition" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS) -Wl,--emit-relocs" ../../../configure --prefix=/usr --enable-optimizations --with-lto --disable-debugging-features
+	cd pyston/build/cpython_opt_build; PROFILE_TASK="$(PROFILE_TASK)" CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -fno-reorder-blocks-and-partition" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS) -Wl,--emit-relocs" ../../../configure --prefix=/usr --enable-optimizations --with-lto --disable-debugging-features --enable-configure
 # We have to --disable-debugging-features for consistency with the bc build
 # If we had a separate bc-dbg build then we could change this
 pyston/build/cpython_dbg_build/Makefile: $(MAKEFILE_DEPENDENCIES)
 	mkdir -p pyston/build/cpython_dbg_build
-	cd pyston/build/cpython_dbg_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -fno-reorder-blocks-and-partition" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS) -Wl,--emit-relocs" ../../../configure --prefix=/usr --with-pydebug --disable-debugging-features
+	cd pyston/build/cpython_dbg_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS) -fno-reorder-blocks-and-partition" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS) -Wl,--emit-relocs" ../../../configure --prefix=/usr --with-pydebug --disable-debugging-features --enable-configure
 pyston/build/cpython_stock_build/Makefile: $(MAKEFILE_DEPENDENCIES)
 	mkdir -p pyston/build/cpython_stock_build
-	cd pyston/build/cpython_stock_build; PROFILE_TASK="$(PROFILE_TASK) || true" CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS)" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --enable-optimizations --with-lto --disable-pyston
+	cd pyston/build/cpython_stock_build; PROFILE_TASK="$(PROFILE_TASK) || true" CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS)" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --enable-optimizations --with-lto --disable-pyston --enable-configure
 pyston/build/cpython_stockunopt_build/Makefile: $(MAKEFILE_DEPENDENCIES)
 	mkdir -p pyston/build/cpython_stockunopt_build
-	cd pyston/build/cpython_stockunopt_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS)" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --disable-pyston
+	cd pyston/build/cpython_stockunopt_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS)" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --disable-pyston --enable-configure
 pyston/build/cpython_stockdbg_build/Makefile: $(MAKEFILE_DEPENDENCIES)
 	mkdir -p pyston/build/cpython_stockdbg_build
-	cd pyston/build/cpython_stockdbg_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS)" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --disable-pyston --with-pydebug
+	cd pyston/build/cpython_stockdbg_build; CC=gcc CFLAGS_NODIST="$(CPYTHON_EXTRA_CFLAGS)" LDFLAGS_NODIST="$(CPYTHON_EXTRA_LDFLAGS)" ../../../configure --prefix=/usr --disable-pyston --with-pydebug --enable-configure
 
 pyston/build/cpython_bc_build/pyston: pyston/build/cpython_bc_build/Makefile $(filter-out $(wildcard Python/aot*.c),$(wildcard */*.c)) $(wildcard */*.h)
 	cd pyston/build/cpython_bc_build; WRAPPER_REALCC=$(realpath $(CLANG)) WRAPPER_OUTPUT_PREFIX=../cpython_bc $(MAKE)

--- a/configure
+++ b/configure
@@ -806,6 +806,7 @@ SHELL'
 ac_subst_files=''
 ac_user_opts='
 enable_option_checking
+enable_configure
 enable_universalsdk
 with_universal_archs
 with_framework_name
@@ -1487,6 +1488,8 @@ Optional Features:
   --disable-option-checking  ignore unrecognized --enable/--with options
   --disable-FEATURE       do not include FEATURE (same as --enable-FEATURE=no)
   --enable-FEATURE[=ARG]  include FEATURE [ARG=yes]
+  --enable-configure      Safety check because users should not call configure
+                          themselves in Pyston
   --enable-universalsdk[=SDKDIR]
                           Build fat binary against Mac OS X SDK
   --enable-framework[=INSTALLDIR]
@@ -2728,7 +2731,24 @@ if test "$srcdir" != . -a "$srcdir" != "$(pwd)"; then
     BASECPPFLAGS="-IObjects -IInclude -IPython"
 else
     BASECPPFLAGS=""
-    as_fn_error $? "pyston does not support in-source builds, please run configure from a separate directory" "$LINENO" 5
+fi
+
+# Check whether --enable-configure was given.
+if test "${enable_configure+set}" = set; then :
+  enableval=$enable_configure;
+	case $enableval in
+	yes)
+	esac
+	case $enableval in
+	no)
+        as_fn_error $? "Please build Pyston from the top-level Makefile, not by calling configure" "$LINENO" 5
+	esac
+
+
+else
+
+    as_fn_error $? "Please build Pyston from the top-level Makefile, not by calling configure" "$LINENO" 5
+
 fi
 
 

--- a/configure.ac
+++ b/configure.ac
@@ -27,9 +27,22 @@ if test "$srcdir" != . -a "$srcdir" != "$(pwd)"; then
     BASECPPFLAGS="-IObjects -IInclude -IPython"
 else
     BASECPPFLAGS=""
-    AC_MSG_ERROR([pyston does not support in-source builds, please run configure from a separate directory])
 fi
 
+AC_ARG_ENABLE(configure,
+	AS_HELP_STRING([--enable-configure], [Safety check because users should not call configure themselves in Pyston]),
+[
+	case $enableval in
+	yes)
+	esac
+	case $enableval in
+	no)
+        AC_MSG_ERROR([Please build Pyston from the top-level Makefile, not by calling configure])
+	esac
+
+],[
+    AC_MSG_ERROR([Please build Pyston from the top-level Makefile, not by calling configure])
+])
 AC_SUBST(GITVERSION)
 AC_SUBST(GITTAG)
 AC_SUBST(GITBRANCH)

